pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go

```diff
@@ -123,51 +133,112 @@ func (cs *cesSubscriber) OnUpdate(oldCES, newCES *cilium_v2a1.CiliumEndpointSlic
 // and calls endpointDeleted only for remoteNode CEPs.
 func (cs *cesSubscriber) OnDelete(ces *cilium_v2a1.CiliumEndpointSlice) {
 	for i, ep := range ces.Endpoints {
+		CEPName := ces.Namespace + "/" + ep.Name
 		log.WithFields(logrus.Fields{
 			"CESName": ces.GetName(),
-			"CEPName": ep.Name,
+			"CEPName": CEPName,
 		}).Debug("CES deleted, calling endpointDeleted")
 		c := k8s.ConvertCoreCiliumEndpointToTypesCiliumEndpoint(&ces.Endpoints[i], ces.Namespace)
 		// LocalNode already deleted the CEP.
 		// Hence, skip processing endpointDeleted for localNode CEPs.
-		if p := cs.kWatcher.endpointManager.LookupPodName(k8sUtils.GetObjNamespaceName(c)); p != nil {
+		if p := cs.epCache.LookupPodName(k8sUtils.GetObjNamespaceName(c)); p != nil {
 			continue
 		}
 		// Delete CEP if and only if that CEP is owned by a CES, that was used during CES updated.
 		// Delete CEP only if there is match in CEPToCES map and also delete CEPName in CEPToCES map.
-		if cesName := cepMap.getCESName(ces.Namespace + "/" + ep.Name); cesName == ces.GetName() {
-			cs.kWatcher.endpointDeleted(c)
-			cepMap.deleteCEP(ep.Name)
+		cs.deleteCEPfromCES(CEPName, ces.GetName(), c)
+	}
+}
+
+// deleteCEP deletes the CEP and CES from the map.
+// If this was last CES for the CEP it triggers endpointDeleted.
+// If this was used CES for the CEP it picks other CES and triggers endpointUpdated.
+func (cs *cesSubscriber) deleteCEPfromCES(CEPName, CESName string, c *types.CiliumEndpoint) {
+	needUpdate := cs.cepMap.currentCES[CEPName] == CESName
```

Comment:
+1, probably create a separate method this?

Reply:
Good catch, thanks. Moved locking to higher methods.

Reply:
I covered also 
cs.epWatcher.endpointDeleted and cs.epWatcher.endpointUpdated with the mutex.

In case of race condition it would be possible that two updates to the same CEP are processed in one order but epWatcher calls are called in the other.

Reply:
lgtm